lib/1.4/utility.dml

```diff
@@ -1393,6 +1395,141 @@ template map_target is (connect, _qname) {
     }
 }
 
+/**
+   ### ram
+
+   This template can be instantiated on a `port`, `device`, `bank` or
+   `subdevice` object for an efficient implementation of a private RAM
+   area. This is implemented by automatically creating a separate Simics object
+   of class `ram`, with an image attached to it, and using the `direct_memory`
+   interface to access the image efficiently. This gives a small memory
+   footprint and efficient checkpointing; this can make a visible impact for
+   internal memories that are several megabytes in size.
+
+   Requires a single integer parameter `size`, which must be set to a multiple of 8192.
+
+   Provides the following methods:
+ */
+typedef struct {
+    direct_memory_handle_t handle;
+    uint8 *data;
+} _ram_page_t;
+template ram {
+    param size;
+    connect ram is (init_as_subobj, destroy) {
```

Comment:
A better design is probably to create a template for a private RAM connect, without  DMI support, and then create a separate template which adds DMI magic to all private RAMs within a bank/subdev/port

Reply:
Oh, this adds some pain: the use of get_user_data requires that you have one separate direct_memory_update ports for each RAM; if we want a shared impl, then we instead need to keep a private hashtable from handles to pages.